### DISCUSSION TITLE
Trimming QEMU from Scratch

### DISCUSSION THREAD

**Discussion by zevorn** [zevorn/blog#10] [blog]
To make it easier to study QEMU's TCG module, I tried trimming the QEMU source tree from scratch and keeping only the TCG-related modules. By pruning the source tree, you can gain a deeper understanding of QEMU's code organization and build a solid foundation for learning it.

> Base QEMU version: 8.2.0

## QEMU build system

QEMU uses Meson + Ninja for builds (Make is also supported). Its scripting style is similar to Python and is highly readable, with fast compilation and build speed. In addition to Meson, you also need to pay attention to QEMU's ./configure script and Kconfig scripts: the former is the top-level configuration file, while the latter contains the source paths for specific builds (used together with meson.build).

So the first step in the trimming effort was to understand all of ./configure's build options. The main idea here was to keep only the user-mode pieces related to TCG, GDB remote debugging, user interaction, plugins, and other core components, and remove all code related to system mode. This can be done by modifying ./configure:

```shell
system="no"
linux_user="yes"
```

The next step is to remove the unneeded code gradually according to meson.build. The following sections describe some of the modules that were removed.

**Comment by xtbing** [comment]
博主，感觉你这篇文章没写完呢，是不是还有连载，想请教下qemu裁剪相关的问题？

**Reply by zevorn** [reply]
可以微信关注 “GTC慢跑计划” 公众号，这是我最近在做的一个 QEMU 相关的课程，可以提供指导。
